piptools/scripts/compile.py

```diff
@@ -51,7 +54,7 @@
               help='Try to upgrade all dependencies to their latest versions')
 @click.option('-P', '--upgrade-package', 'upgrade_packages', nargs=1, multiple=True,
               help="Specify particular packages to upgrade.")
-@click.option('-o', '--output-file', nargs=1, type=str, default=None,
+@click.option('-o', '--output-file', nargs=1, default=None, type=click.File('w+b', atomic=True, lazy=True),
               help=('Output file name. Required if more than one input file is given. '
                     'Will be derived from input file otherwise.'))
```

Comment:
IIRC `click.File` creates the file eagerly already, i.e. it might overwrite/truncate an existing one, also if pip-compile would error out before actually writing it.

So I would rather keep this as a string if possible.

Reply:
Pay attention to the `lazy=True` flag:
>     The `lazy` flag controls if the file should be opened immediately or upon
>     first IO. The default is to be non-lazy for standard input and output
>     streams as well as files opened for reading, `lazy` otherwise. When opening a
>     file lazily for reading, it is still opened temporarily for validation, but
>     will not be held open until first IO. lazy is mainly useful when opening
>     for writing to avoid creating the file until it is needed.
And there is also an atomic mode, which means that file wouldn't be overwritten unless its closed.

Reply:
Oh, great then!